### PR TITLE
Feature/saml based firewall

### PR DESCRIPTION
**Pull request by arothuis**
[Ticket 106201244](https://www.pivotaltracker.com/story/show/106201244)

Done:
* Configure SurfnetSamlBundle
* Setup a SAML-based firewall based on Stepup-SelfService and Stepup-Gateway
* Check configuration with EngineBlock

Issues:
* Our current dev vm seems to cause some trouble (permission denied or path not found) when using (native) php sessions. Fixed by #3.

> `Warning: SessionHandler::read(): open(/var/lib/php/session/profile.vm.openconext.org/sess_96rpcah8ba6i2k0drcu04i5qi7, O_RDWR) failed: No such file or directory (2)`

### CODE REVIEW

**Comment by arothuis** [comment]
@rjkip

File: src/OpenConext/ProfileBundle/Saml/StateHandler.php [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        $this->session->migrate();
+    }
+
+    /**
+     * @param string $key
+     * @param mixed $value Any scalar
+     */
+    protected function set($key, $value)
+    {
+        $this->attributeBag->set(self::SESSION_PATH . $key, $value);
```
Comment (rjkip) [inline src/OpenConext/ProfileBundle/Saml/StateHandler.php]: `SESSION_PATH` is not needed when working with a namespaced attribute bag. Might as well replace all calls to `$this->set()` with `$this->attributeBag->set()`.

File: src/OpenConext/ProfileBundle/Security/Authentication/Entity/User.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+     */
+    public $email;
+
+    /**
+     * @var string
+     */
+    public $commonName;
+
+    /**
+     * Using toString in order to comply with AbstractToken's setUser method.
```
Comment (rjkip) [inline src/OpenConext/ProfileBundle/Security/Authentication/Entity/User.php]: What about

```diff
-setUser method.
+setUser method, which uses the string representation to detect changes in the user object.
```

File: src/OpenConext/ProfileBundle/Security/Authentication/Provider/SamlProvider.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+class SamlProvider implements AuthenticationProviderInterface
+{
+    /**
+     * @var \Surfnet\SamlBundle\SAML2\Attribute\AttributeDictionary
+     */
+    private $attributeDictionary;
+
+    public function __construct(
+        AttributeDictionary $attributeDictionary
```
Comment (rjkip) [inline src/OpenConext/ProfileBundle/Security/Authentication/Provider/SamlProvider.php]: Nit: move to one line.

File: src/OpenConext/ProfileBundle/Security/Authentication/Provider/SamlProvider.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        $user->email = $translatedAssertion->getAttribute('mail');
+        $user->commonName = $translatedAssertion->getAttribute('commonName');
+
+        $authenticatedToken = new SamlToken(['ROLE_USER']);
+
+        $authenticatedToken->setUser($user);
+
+        return $authenticatedToken;
+    }
+    public function supports(TokenInterface $token)
```
Comment (rjkip) [inline src/OpenConext/ProfileBundle/Security/Authentication/Provider/SamlProvider.php]: Nit: prepend newline

File: src/OpenConext/ProfileBundle/Security/Authentication/Provider/SamlProvider.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        AttributeDictionary $attributeDictionary
+    ) {
+        $this->attributeDictionary = $attributeDictionary;
+    }
+
+    /**
+     * @param  TokenInterface $token
+     * @return TokenInterface|void
+     */
+    public function authenticate(TokenInterface $token)
```
Comment (rjkip) [inline src/OpenConext/ProfileBundle/Security/Authentication/Provider/SamlProvider.php]: Is this method part of the interface? Then docblocks are discouraged (read: remove :wink:).

File: src/OpenConext/ProfileBundle/Security/Authentication/SamlInteractionProvider.php [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+     * @return \Symfony\Component\HttpFoundation\RedirectResponse
+     */
+    public function initiateSamlRequest()
+    {
+        $authnRequest = AuthnRequestFactory::createNewRequest(
+            $this->serviceProvider,
+            $this->identityProvider
+        );
+        $this->stateHandler->setRequestId($authnRequest->getRequestId());
+        return $this->redirectBinding->createRedirectResponseFor($authnRequest);
```
Comment (rjkip) [inline src/OpenConext/ProfileBundle/Security/Authentication/SamlInteractionProvider.php]: We typically insert newlines before return statements.

File: src/OpenConext/ProfileBundle/Security/Authentication/Token/SamlToken.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        parent::__construct($roles);
+        $this->setAuthenticated(count($roles));
+    }
+    /**
+     * Returns the user credentials.
+     *
+     * @return mixed The user credentials
+     */
+    public function getCredentials()
```
Comment (rjkip) [inline src/OpenConext/ProfileBundle/Security/Authentication/Token/SamlToken.php]: Doc blocks are inherited from parent.

**Comment by rjkip** [comment]
A lot of newlines seem to be missing. After that, and assuming you performed an actual SAML authentication, :+1:

**Comment by rjkip** [comment]
:+1:

**Pull request merged**